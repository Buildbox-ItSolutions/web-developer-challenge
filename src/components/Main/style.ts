import styled from "styled-components";

export const StyledMain = styled.main`
  display: flex;
  align-items: center;
  justify-content: center;
  padding-top: 32px;
  padding-bottom: 64px;
  flex-direction: column;
  gap: 64px;
`;
