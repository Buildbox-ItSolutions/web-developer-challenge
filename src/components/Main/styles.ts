import styled from 'styled-components'

export const Wrapper = styled.main`
  background-color: #06092b;
  color: #fff;
  width: 100%;
  height: 100%;
  padding: 3rem;
  text-align: center;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  overflow-y: scroll;
`
export const Post = styled.div``

export const All = styled.div`
  background-color: #343434;
`
