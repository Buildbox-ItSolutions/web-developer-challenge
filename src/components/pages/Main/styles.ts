import styled from 'styled-components';

export const Container = styled.div`
  animation: transitionIn 1s;
  @keyframes transitionIn {
    from {
      opacity: 0;
    }

    to {
      opacity: 1;
    }
  }

  width: 100vw;

  main {
    padding: 10vh 10vw;

    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
  }

  @media (max-width: 600px) {
    main {
      padding: 5vw 0;
    }
  }
`;
