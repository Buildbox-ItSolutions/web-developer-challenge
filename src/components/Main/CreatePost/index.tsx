/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @next/next/no-img-element */

import { useContext, useRef, useState } from 'react'
import { v4 as uuidv4 } from 'uuid'
import { ToastContainer, toast } from 'react-toastify'
import 'react-toastify/dist/ReactToastify.css'

import PostsContext from '../../../shared/contexts/posts'
import { getBase64 } from '../../../shared/utils/getBase64'
import { CreatePostContainer } from './style'

export default function CreatePost() {
  const { posts, setPosts } = useContext(PostsContext)

  const [base64, setBase64] = useState<string>('')

  const inputNameRef = useRef<HTMLInputElement>(null)
  const inputMessageRef = useRef<HTMLTextAreaElement>(null)

  function cleanForm() {
    if (inputNameRef.current && inputMessageRef.current) {
      setBase64('')
      inputNameRef.current.value = ''
      inputMessageRef.current.value = ''
    }
  }

  async function handleSubmit(event: React.FormEvent) {
    event.preventDefault()

    if (
      inputNameRef.current &&
      inputMessageRef.current &&
      inputNameRef.current.value !== '' &&
      inputMessageRef.current.value !== '' &&
      base64 !== ''
    ) {
      const newPost = {
        id: uuidv4(),
        image: base64,
        name: inputNameRef.current.value,
        message: inputMessageRef.current.value,
      }

      setPosts([...posts, newPost])

      cleanForm()

      toast.success('Publicação enviada com sucesso!', {
        position: toast.POSITION.TOP_CENTER,
        autoClose: 2000,
      })
    } else {
      if (base64 === '') {
        toast.error('Adicione uma imagem à publicação!', {
          position: toast.POSITION.TOP_CENTER,
          autoClose: 2000,
        })
      }
      if (inputNameRef.current && inputNameRef.current.value === '') {
        toast.error('Preencha seu nome na publicação!', {
          position: toast.POSITION.TOP_CENTER,
          autoClose: 2000,
        })
      }

      if (inputMessageRef.current && inputMessageRef.current.value === '') {
        toast.error('Escreva uma mensagem na publicação!', {
          position: toast.POSITION.TOP_CENTER,
          autoClose: 2000,
        })
      }
    }
  }

  async function onChangeFileInput(file: File) {
    const newBase64 = await getBase64(file)

    setBase64(newBase64 as string)
  }

  return (
    <CreatePostContainer>
      <div className="card">
        <form
          onSubmit={(event) => {
            handleSubmit(event)
          }}
        >
          <div className="selectImage">
            {base64 === '' ? (
              <label htmlFor="imageFileInput">
                <img src="/image.png" alt="image icon" />
              </label>
            ) : (
              <label htmlFor="imageFileInput">
                <img
                  src={base64}
                  alt="selected image"
                  className="selectedImage"
                />
              </label>
            )}
            <input
              id="imageFileInput"
              type="file"
              onClick={(e) => ((e.target as HTMLInputElement).value = '')}
              onChange={(e) => {
                if (e.target.files) {
                  onChangeFileInput(e.target.files[0])
                }
              }}
              accept=".png, .jpg, .jpeg"
            />
            {base64 !== '' ? (
              <img
                src="/trash.png"
                alt="trash icon"
                className="trashIcon"
                onClick={() => setBase64('')}
              />
            ) : (
              ''
            )}
          </div>
          <input
            className="inputName"
            ref={inputNameRef}
            type="text"
            name="name"
            placeholder="Digite seu nome"
          />
          <textarea ref={inputMessageRef} placeholder="Mensagem" />
          <div className="buttonsArea">
            <button className="descartar" onClick={cleanForm} type="button">
              Descartar
            </button>
            <button className="publicar" type="submit">
              Publicar
            </button>
          </div>
        </form>
      </div>
      <ToastContainer limit={3} />
    </CreatePostContainer>
  )
}
