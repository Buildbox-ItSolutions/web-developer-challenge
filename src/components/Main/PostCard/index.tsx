import Image from 'next/image'
import { useContext } from 'react'
import { toast, ToastContainer } from 'react-toastify'
import PostsContext from '../../../shared/contexts/posts'
import { PostProps } from '../../../shared/interfaces/post'
import { PostCardContainer } from './style'

interface PostCardProps {
  post: PostProps
}

export default function PostCard({ post }: PostCardProps) {
  const { posts, setPosts } = useContext(PostsContext)

  const loaderImageProp = ({ src }: { src: string }) => {
    return src
  }

  function deletePost() {
    const newPosts = posts.filter((actualPost) => {
      return actualPost.id !== post.id
    })

    setPosts(newPosts)

    toast.error('Publicação excluida com sucesso!', {
      position: toast.POSITION.TOP_CENTER,
      autoClose: 2000,
    })
  }

  return (
    <PostCardContainer>
      <div className="closeButtonArea" onClick={deletePost}>
        <Image
          src="/delete.png"
          alt="close button"
          width="20px"
          height="20px"
        />
      </div>
      <div className="cardArea">
        <div className="imageArea">
          <Image
            src={post.image}
            loader={loaderImageProp}
            alt="Post image"
            width="88px"
            height="88px"
            className="imagePost"
          />
        </div>
        <div className="textArea">
          <h1 className="message">{post.message}</h1>
          <div>
            <h6>Enviado por</h6>
            <h4 className="name">{post.name}</h4>
          </div>
        </div>
      </div>
      <ToastContainer limit={3} />
    </PostCardContainer>
  )
}
