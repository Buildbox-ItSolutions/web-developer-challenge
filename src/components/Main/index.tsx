import React from "react";
import { StyledMain } from "./style";

interface MainProps {
  children: React.ReactNode;
}

export const Main = ({ children }: MainProps) => {
  return <StyledMain>{children}</StyledMain>;
};
