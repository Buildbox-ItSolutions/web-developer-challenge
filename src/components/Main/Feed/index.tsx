import { useContext } from 'react'

import PostsContext from '../../../shared/contexts/posts'
import Spinner from '../../Spinner'
import PostCard from '../PostCard'
import { FeedContainer } from './style'

export default function Feed() {
  const { posts } = useContext(PostsContext)

  if (!posts) {
    return (
      <FeedContainer>
        <div className="titleArea">
          <h1>Feed</h1>
        </div>
        <div style={{ margin: '2rem 0' }}>
          <Spinner />
        </div>
      </FeedContainer>
    )
  } else {
    return (
      <FeedContainer>
        <div className="titleArea">
          <h1 className="title">Feed</h1>
        </div>
        {posts.length === 0 ? (
          <h1 className="noContent">Nenhuma publicação encontrada...</h1>
        ) : (
          posts.map((post) => {
            return <PostCard key={post.id} post={post} />
          })
        )}
      </FeedContainer>
    )
  }
}
