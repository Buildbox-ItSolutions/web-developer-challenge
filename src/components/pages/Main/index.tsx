import React from 'react';
import { useSelector } from 'react-redux';

import { Header } from 'components/atoms';
import { CreatePost, PostItem } from 'components/molecules';

import { RootState } from 'store/ducks/rootReducer';
import { Container } from './styles';

const Main: React.FC = () => {
  const posts = Array.from(useSelector((state: RootState) => state.posts));

  return (
    <>
      <Container>
        <Header />
        <main>
          <CreatePost />
          {posts.map(post => (
            <PostItem postData={post} key={post.id} />
          ))}
        </main>
      </Container>
    </>
  );
};

export default Main;
