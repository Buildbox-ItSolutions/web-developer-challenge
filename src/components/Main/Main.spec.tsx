import React from 'react'
import { render } from '@testing-library/react'
import Main from '.'

describe('<Main />', () => {
  it('dummy test', () => {
    render(<Main />)
  })

  it('should match snapshot', () => {
    const { container } = render(<Main />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
