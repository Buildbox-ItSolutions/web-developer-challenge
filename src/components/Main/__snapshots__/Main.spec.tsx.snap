// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main /> should match snapshot 1`] = `
.c14 {
  width: 88px;
  height: 88px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin-right: 32px;
  border-radius: 36px;
  background-color: #5f5f5f;
}

.c14 img {
  width: 100%;
  object-fit: contain;
}

.c11 {
  width: 100%;
  height: 185px;
  max-width: 516px;
  position: relative;
  padding: 32px;
  margin-bottom: 24px;
  border-radius: 3px;
  border: solid 1px #3b3b3b;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #313131;
}

.c12 {
  width: 20px;
  height: 20px;
  border: none;
  cursor: pointer;
  -webkit-transition: ease-in-out 0.3s;
  transition: ease-in-out 0.3s;
  top: 10px;
  right: 15px;
  position: absolute;
  background-color: transparent;
}

.c12:hover {
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c13 {
  width: 100%;
  margin-top: 32px;
  margin-bottom: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c15 {
  font-size: 16px;
  color: #9f9f9f;
}

.c16 {
  margin-bottom: 32px;
}

.c16 > span {
  font-size: 12px;
  font-weight: 500;
  margin-bottom: 4px;
  display: inline-block;
  color: #5f5f5f;
}

.c16 > p {
  font-size: 14px;
  color: #7a7a7a;
}

.c9 {
  width: 60px;
  padding: 16px;
  position: relative;
  cursor: pointer;
  -webkit-transition: ease-in-out 0.3s;
  transition: ease-in-out 0.3s;
  font-size: 14px;
  font-weight: normal;
  background-color: transparent;
  color: #5f5f5f;
}

.c9::after {
  content: '';
  width: 100%;
  height: 2px;
  display: block;
  top: 32px;
  position: absolute;
  background-color: #5f5f5f;
}

.c9:hover {
  opacity: 0.9;
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c10 {
  width: 98px;
  height: 41px;
  border-radius: 8px;
  padding: 0;
  cursor: pointer;
  -webkit-transition: ease-in-out 0.3s;
  transition: ease-in-out 0.3s;
  text-align: center;
  color: #ffffff;
  background-color: #71bb00;
}

.c10:hover {
  opacity: 0.9;
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c2 {
  width: 100%;
  max-width: 516px;
  margin: 24px auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 24px;
  border-radius: 3px;
  border: solid 1px #3b3b3b;
  background-color: #313131;
}

.c3 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c4 {
  width: 88px;
  height: 88px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  background-color: transparent;
  border-radius: 30px;
  border: 1px solid #494949;
}

.c4 img {
  object-fit: contain;
  max-width: 100%;
}

.c4 .inputFile {
  display: none;
}

.c5 {
  width: 20px;
  margin-left: 24px;
  border: none;
  background-color: transparent;
  cursor: pointer;
  -webkit-transition: ease-in-out 0.3s;
  transition: ease-in-out 0.3s;
}

.c5:hover {
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c6 {
  width: 100%;
  height: 40px;
  max-width: 468px;
  padding: 16px;
  margin: 16px 0 8px;
  font-size: 14px;
  border-radius: 8px;
  color: #ffffff;
  background-color: #494949;
}

.c7 {
  width: 100%;
  height: 80px;
  max-width: 468px;
  margin: 8px;
  padding: 12px 16px;
  font-size: 14px;
  font-family: 'Roboto',sans-serif;
  border: none;
  outline: none;
  resize: none;
  border-radius: 8px;
  color: #ffffff;
  background-color: #494949;
}

.c8 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c8 > button:nth-child(1) {
  margin-right: 32px;
}

.c1 {
  width: 100%;
  padding: 24px 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  box-shadow: 2px 2px 2px rgba(0,0,0,0.3);
  background-color: #2b2b2b;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<main
  class="c0"
>
  <header
    class="c1"
  >
    <img
      alt="Logo escrito 'buildbox web challenge'"
      src="bx-logo.png"
    />
  </header>
  <form
    class="c2"
  >
    <div
      class="c3"
    >
      <label
        class="c4"
        for="photo"
      >
        <img
          alt="Icone de imagem"
          src="image.png"
        />
        <input
          class="inputFile"
          id="photo"
          name="photo"
          type="file"
        />
      </label>
      <button
        class="c5"
      >
        <svg>
          trash.svg
        </svg>
      </button>
    </div>
    <input
      class="c6"
      name="name"
      placeholder="Nome"
      required=""
      type="text"
      value=""
    />
    <textarea
      class="c7"
      name="menssagem"
      placeholder="Mensagem"
      required=""
    />
    <div
      class="c8"
    >
      <button
        class="c9"
      >
        Descartar
      </button>
      <button
        class="c10"
        type="submit"
      >
        Publicar
      </button>
    </div>
  </form>
  <div
    class=""
  >
    <div
      class="c11"
    >
      <button
        class="c12"
      >
        <svg>
          delete.svg
        </svg>
      </button>
      <div
        class="c13"
      >
        <div
          class="c14"
        >
          <img
            alt="Lorem ipsum dolor sit amet"
            src="/image/image.png"
          />
        </div>
        <p
          class="c15"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla mattis ligula vel velit scelerisque iaculis. Nam mattis justo id orci commodo, eu tempus purus cursus.
        </p>
      </div>
      <div
        class="c16"
      >
        <span>
          Enviado por
        </span>
        <p>
          Manuela Oliveira
        </p>
      </div>
    </div>
    <div
      class="c11"
    >
      <button
        class="c12"
      >
        <svg>
          delete.svg
        </svg>
      </button>
      <div
        class="c13"
      >
        <div
          class="c14"
        >
          <img
            alt="Lorem ipsum dolor sit amet"
            src="/image/image.png"
          />
        </div>
        <p
          class="c15"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla mattis ligula vel velit scelerisque iaculis. Nam mattis justo id orci commodo, eu tempus purus cursus.
        </p>
      </div>
      <div
        class="c16"
      >
        <span>
          Enviado por
        </span>
        <p>
          Manuela Oliveira
        </p>
      </div>
    </div>
  </div>
</main>
`;
